src/pcp.rs

```diff
@@ -0,0 +1,596 @@
+        let mut p_vals = vec![F::zero(); size];
+        discrete_fourier_transform(&mut p_vals, &pf.data[l..], size)?;
+
+        let step = (1 << (log2(p as u128) - log2(m as u128))) as usize;
+        Ok(QueryShimGadget {
+            inner,
+            f_vals,
+            p_vals,
+            step,
+            ct: 1,
+        })
+    }
+}
+
+impl<'a, F, G> GadgetCallOnly<F> for QueryShimGadget<'a, F, G>
+where
+    F: FieldElement,
+    G: Gadget<F>,
+{
+    fn call(&mut self, inp: &[F]) -> Result<F, PcpError> {
+        for i in 0..inp.len() {
+            self.f_vals[i][self.ct] = inp[i];
+        }
+        let outp = self.p_vals[self.ct * self.step];
+        self.ct += 1;
+        Ok(outp)
+    }
+
+    fn call_in_len(&self) -> usize {
+        self.inner.call_in_len()
+    }
+}
+
+/// The output of `query`, the verifier message generated for a proof.
+#[derive(Debug)]
+pub struct Verifier<F: FieldElement> {
+    v: F,           // Output of the validity circuit
+    p_at_r: F,      // The proof polynomial evluated at `r`
+    f_at_r: Vec<F>, // The intermediate proof polynomials evaluated at `r`
+}
+
+/// Returns the verifier corresponding to a sequence of verifier shares.
+pub fn collect<F: FieldElement>(vf_shares: &[Verifier<F>]) -> Result<Verifier<F>, PcpError> {
+    if vf_shares.len() == 0 {
+        return Err(PcpError::CollectInLen);
+    }
+
+    let l = vf_shares[0].f_at_r.len();
+    let mut vf = Verifier {
+        v: F::zero(),
+        p_at_r: F::zero(),
+        f_at_r: vec![F::zero(); l],
+    };
+
+    for i in 0..vf_shares.len() {
+        if vf_shares[i].f_at_r.len() != l {
+            return Err(PcpError::CollectGadgetInLenMismatch);
+        }
+
+        vf.v += vf_shares[i].v;
+        vf.p_at_r += vf_shares[i].p_at_r;
+        for j in 0..l {
+            vf.f_at_r[j] += vf_shares[i].f_at_r[j];
+        }
+    }
+
+    Ok(vf)
+}
+
+/// Decide if the input (or input share) is valid using the given verifier.
+pub fn decide<F, G, T>(x: &T, vf: &Verifier<F>) -> Result<bool, PcpError>
+where
+    F: FieldElement,
+    G: Gadget<F>,
+    T: Type<F, G>,
+{
+    let e = x.gadget(0).call(&vf.f_at_r)?;
+    if e == vf.p_at_r && vf.v == F::zero() {
+        Ok(true)
+    } else {
+        Ok(false)
+    }
+}
+
+// Returns a polynomial `p` for which `p(x) = 0` if in `[start, end)` and `p(x) != 0` otherwise.
+fn poly_range_check<F: FieldElement>(start: usize, end: usize) -> Vec<F> {
+    let mut p = vec![F::one()];
+    let mut q = [F::zero(), F::one()];
+    for i in start..end {
+        q[0] = -F::from(F::Integer::try_from(i).unwrap());
+        p = poly_mul(&p, &q);
+    }
+    p
+}
+
+fn rand_vec<F: FieldElement>(len: usize) -> Vec<F> {
+    let mut outp: Vec<F> = Vec::with_capacity(len);
+    for _ in 0..len {
+        outp.push(F::rand());
+    }
+    outp
+}
+
+/// Outputs an additive secret sharing of the input.
+pub fn split<F: FieldElement>(inp: &[F], num_shares: usize) -> Vec<Vec<F>> {
```

Comment:
This also feels like it belongs in `field.rs`.

Reply:
Agreed. Moved.